### PR TITLE
Fix #335 - Correct the behaviour of Push/Pull button.

### LINKED ISSUE #335: Activeness of pull/push button.
#### Software versions 
Briefcase v1.9.x, probably preview versions too, Java v1.8.x, operating system: All

#### Problem description
I have noticed differences between activeness of pull, push and export button.

#### Steps to reproduce the problem
1.Run briefcase and do not connect to the aggregate server. - Pull button is active, push button is active too.
2. Connect to the aggregate server on pull page - pull button is disabled.
3. Connect to the aggregate server on push page - push button is active.
4. Click on push button, nothing happens.
5. Go back to pull page and you can see that pull button become active. 

#### Expected behavior
I think that pull and push button should act in the same way as an export button which is active only choosing export direction and form from export list. Means they should be active only after connection to the aggregate server and choosing form from the checklist.

#### Other information 
It is not high priority issue but it would be nice to unify all buttons.

### PR DESCRIPTION
**Pull request by mayank8318**
Closes #335 

#### What has been done to verify that this works as intended?
Ran the software, and it worked as intended. The push button is active when there is a destination and at least one form is selected. The pull button is active when at least one form is selected. 

#### Why is this the best possible solution? Were any other approaches considered?
Instead of setting the button status based on "transferring/not transferring" I am using the number of selected forms to check if the transfer button should be active or not.

#### Are there any risks to merging this code? If so, what are they?
Haven't tested this code via tests.

### CODE REVIEW

**Comment by dcbriccetti** [comment]
I started to make review comments, but ended up modifying your change to remove what I saw as a small separation of concerns problem. Here’s what I did:
https://github.com/dcbriccetti/briefcase/commit/452a18ddc6e3e3dcaf3c63a90aaafb16fea9d7b6
What do you think? Would you like to cherry pick that commit into your branch and continue, or should I create a new PR that builds on your work?
@ggalmazor, I’d be interested in whether you think my changes represent an improvement.

**Comment by ggalmazor** [comment]
Thanks, @dcbriccetti! I like that the change you're proposing goes in the same direction of what we're doing on the Export tab.

There, the controller `ExportPanel` class owns an instance of `ExportConfiguration` and buttons are enabled/disabled regarding its state of emptiness/validness. All children components use this conf object but they don't make changes to it. Having the `PullTransferPanel` own a `ServerSelectedObservable` instance which is passed to children components is just the same thing.

This concept of state ownership is central to the component design rules we're applying on the Export tab, and which should be extended to other tabs. 

Nevertheless, I'm not so sure about introducing `Observable` because we don't use it anywhere else. Instead, we inject `Consumer` and `Runnable` callbacks, which don't require creating specific classes for each thing you need to observe. 

Moreover, when we review parent-child and cross-app component communication in the near future all this is subject to change. Then, the least amount of moving parts we have, the better.

Could this be achieved by passing callbacks, like on the Export tab?

**Comment by dcbriccetti** [comment]
@mayank8318, do you understand what @ggalmazor is recommending above, and want to continue implementing it?

**Comment by mayank8318** [comment]
@dcbriccetti I will be busy this week due to college quizzes. So you can implement it.
Thanks.

**Comment by ggalmazor** [comment]
OK @mayank8318, I hope you luck with your quizzes this week! :) 

I think that your PR it's fine overall and it just needs some changes to break those coupling problems @dcbriccetti mentioned. I will make my review and, if you want, we can continue with it next week.

File: src/org/opendatakit/briefcase/ui/FormTransferTable.java [hunk 1]
```diff
@@ -316,8 +324,9 @@ public void fireStatusChange(FormStatusEvent fse) {
 
   }
 
-  public FormTransferTable(JButton btnSelectOrClearAllForms, FormStatus.TransferType transferType, JButton btnTransfer, JButton btnCancel) {
-    super(new FormTransferTableModel(btnSelectOrClearAllForms, transferType, btnTransfer, btnCancel));
+  public FormTransferTable(JButton btnSelectOrClearAllForms, TransferType transferType, JButton btnTransfer,
```
Comment (ggalmazor) [inline src/org/opendatakit/briefcase/ui/FormTransferTable.java]: Don't receive the text field

File: src/org/opendatakit/briefcase/ui/PullTransferPanel.java [hunk 2]
```diff
@@ -266,7 +266,7 @@ public void actionPerformed(ActionEvent arg0) {
     });
 
     formTransferTable = new FormTransferTable(
-        btnSelectOrClearAllForms, FormStatus.TransferType.GATHER, btnTransfer, btnCancel);
+        btnSelectOrClearAllForms, FormStatus.TransferType.GATHER, btnTransfer, btnCancel, null);
```
Comment (ggalmazor) [inline src/org/opendatakit/briefcase/ui/PullTransferPanel.java]: Remove all the changes on this file and center your attention on `OriginActionListener.actionPerformed(ActionEvent e)`. Once we know that the user has set the source settings, we can call the table's `setSourceSelected(boolean)` method.

File: src/org/opendatakit/briefcase/ui/PushTransferPanel.java [hunk 3]
```diff
@@ -207,7 +209,7 @@ public void actionPerformed(ActionEvent arg0) {
     });
 
     formTransferTable = new FormTransferTable(
-        btnSelectOrClearAllForms, FormStatus.TransferType.UPLOAD, btnTransfer, btnCancel);
+        btnSelectOrClearAllForms, FormStatus.TransferType.UPLOAD, btnTransfer, btnCancel, txtDestinationName);
```
Comment (ggalmazor) [inline src/org/opendatakit/briefcase/ui/PushTransferPanel.java]: Same changes as in `PullTransferPanel`

**Comment by mayank8318** [comment]
@ggalmazor thank you for the good wishes😊. If it is not urgent I would really like to work on it next week.

File: src/org/opendatakit/briefcase/ui/PullTransferPanel.java [hunk 4]
```diff
@@ -429,7 +429,7 @@ private void setActiveTransferState(boolean active) {
       listOriginDataSource.setEnabled(true);
       btnOriginAction.setEnabled(true);
       btnSelectOrClearAllForms.setEnabled(true);
-      btnTransfer.setEnabled(true);
+      btnTransfer.setEnabled(formTransferTable.getSelectedForms().size() != 0);
```
Comment (dcbriccetti) [inline src/org/opendatakit/briefcase/ui/PullTransferPanel.java]: I think we’re building a consensus that isEmpty is preferred.

File: src/org/opendatakit/briefcase/ui/PushTransferPanel.java [hunk 5]
```diff
@@ -169,7 +172,7 @@ public void actionPerformed(ActionEvent e) {
   /**
    * Create the transfer-from-to panel.
    *
-   * @param txtBriefcaseDir
+   * @param terminationFuture
```
Comment (dcbriccetti) [inline src/org/opendatakit/briefcase/ui/PushTransferPanel.java]: I know you didn’t create it, but these @params with no explanations have no value and should be deleted.

Comment (mayank8318) [reply src/org/opendatakit/briefcase/ui/PushTransferPanel.java]: Parameter name was wrong so I just corrected it. Will remove it.

Comment (dcbriccetti) [reply src/org/opendatakit/briefcase/ui/PushTransferPanel.java]: Thanks. What you first did was reasonable.

**Comment by yanokwa** [comment]
You are doing great work, @mayank8318! One suggestion I have is that instead of "manually tested" can you please be more explicit about what you tested? If you don't specify, then reviewers have to assume it's untested and we duplicate a lot of work that way.

**Comment by ggalmazor** [comment]
Thanks @mayank8318! Pull tab works great but I can't get the push button enabled...

**Comment by mayank8318** [comment]
@ggalmazor Did you select the directory as well?

**Comment by ggalmazor** [comment]
This is what I got:

no_push

**Comment by mayank8318** [comment]
@ggalmazor Have a class now, will be back in an hour to solve it:)

**Comment by mayank8318** [comment]
@ggalmazor @daveycrockett the tests are passing locally for me?

**Comment by ggalmazor** [comment]
Hi @mayank8318, don't worry, that test is giving us work today... It's not related to your work.

**Comment by ggalmazor** [comment]
Can you just make some minor change in some file to trigger a new build, please?

**Comment by yanokwa** [comment]
Instead of making a minor change to trigger a build, show the details in the CircleCI dialog at the bottom of each PR, click on Details to go to CircleCI then in the upper right hand corner, hit rebuild.

**Comment by ggalmazor** [comment]
I think you need to have some kind of permission that we don't have to do a rebuild...

**Comment by yanokwa** [comment]
Be the change you want to see in the world, @ggalmazor!

**Comment by mayank8318** [comment]
@ggalmazor Sorry about that. Saw a few redundant casts, so removed it. Didn't realise I spoiled the indentation too. Should I remove those changes?

**Comment by ggalmazor** [comment]
No, don't worry! Sooner or later we would have to do that anyway :)

**Comment by dcbriccetti** [comment]
I’m withdrawing from this PR. @ggalmazor’s got it.

**Comment by ggalmazor** [comment]
I think this is ready to be tested. @dcbriccetti, could you approve your review, please?

**Comment by kkrawczyk123** [comment]
Checked on Ubuntu, Mac and on Windows and looks good!

**Comment by kkrawczyk123** [comment]
@opendatakit-bot unlabel "needs testing"
